source/tutorials/nix-language.md

```diff
@@ -0,0 +1,1602 @@
+(indented-strings)=
+### Indented strings
+
+Also known as “multi-line strings”.
+
+Nix language offers convenience syntax for character strings which span multiple lines that have common indentation.
+
+Indented strings are denoted by *double single quotes* (`'' ''`).
+
+Example:
+
+```nix
+''
+multi
+line
+string
+''
+```
+
+    "multi\nline\nstring"
+
+Equal amounts of prepended white space are trimmed from the result.
+
+Example:
+
+```nix
+''
+  one
+   two
+    three
+''
+```
+
+    "one\n two\n  three\n"
+
+<!-- TODO: See [escaping rules](). -->
+
+(antiquotation)=
+### Antiquotation `${ ... }`
+
+Also known as “string interpolation”.
+
+The value of a Nix expression can be inserted into a character string with the dollar-sign and braces (`${ }`).
+
+Example:
+
+```nix
+let
+  name = "Nix";
+in
+"hello ${name}"
+```
+
+    "hello Nix"
+
+Only character strings or values that can be represented as a character string are allowed.
+
+Counter-example:
+
+```nix
+let
+  x = 1;
+in
+"${x} + ${x} = ${x + x}"
+```
+
+    error: cannot coerce an integer to a string
+
+           at «string»:4:2:
+
+                3| in
+                4| "${x} + ${x} = ${x + x}"
+                 |  ^
+                5|
+
+Antiquotation can be arbitrarily nested.
+
+(This can become hard to read, and we recommend to avoid it in practice.)
+
+Example:
+
+```nix
+let
+  a = "no";
+in
+"${a + "${a + " ${a}"}"}"
+```
+
+    "no no no"
+
+
+:::{warning}
+You may encounter strings that use the dollar sign (`$`) before an assigned name, but no braces (`{ }`):
+
+These are *not* antiquotations, but usually denote variables in a shell script.
+
+In such cases, the use of names from the surrounding Nix expression is a coincidence.
+
+Example:
+
+```nix
+let
+  out = "Nix"
+in
+"echo ${out} > $out"
+```
+
+    "echo Nix > $out"
+:::
+
+<!-- TODO: link to escaping rules -->
+
+(functions)=
+## Functions
+
+Functions are everywhere in the Nix language and deserve particular attention.
+
+### Single argument
+
+Functions in Nix language can appear in different forms, but always take exactly one argument.
+Argument and function body are separated by a colon (`:`).
+
+Wherever you see a colon (`:`) in Nix language code:
+- on its left is the function argument
+- on its right is the function body.
+
+Nix functions have no names.
+We say they are anonymous, and call such a function a *lambda*.
+
+Example:
+
+```nix
+x: x + 1
+```
+
+    <LAMBDA>
+
+The `<LAMBDA>` indicates the resulting value is an anonymous function.
+
+We can assign functions a name as to any other value.
+
+Example:
+
+```nix
+let
+  f = x: x + 1
+in f
+```
+
+    <LAMBDA>
+
+### Calling functions
+
+Also known as "function application".
+
+Calling a function with an operand means writing the operand after the function.
+
+Example:
+
+```nix
+(x: x + 1) 1
+```
+
+    2
+
+:::{note}
+Since function and operand are separated by white space, sometimes parantheses (`( )`) are required to distinguish expressions.
```

Comment:
Since the goal of the tutorial is to read the language, not to write it, I'm convinced it doesn't matter much where semicolons go. But I see what you mean, and will think about how to address these issues.

To answer your immediate question: Semicolons are used for three things.
- delimit attributes in a attribute sets
- delimit `let` declarations
- separate source and sink in `with`